0000-crypto-construct-language.md

```diff
@@ -0,0 +1,917 @@
+result is pushed onto the stack.
+
+### Stack Control
+
+```
+/ a -- argument of any type.
+DUP ( a -- a a )
+```
+
+The `DUP` opcode pops the top item, duplicates it and pushes the original and
+its duplicate on the top of the stack.
+
+```
+/ a -- argument of any type.
+POP ( a -- )
+```
+
+The `POP` opcode pops the top item from the stack and forgets about it. This is
+used for throwing away the top of the stack.
+
+### Flow Control
+
+```
+/ b -- boolean argument.
+IF ( b -- ) ELSE FI
+```
+
+The `IF-ELSE-FI` opcode trio and the related `IF-FI` opcode duo are used
+to do conditional branching in CCLang scripts. The `IF` opcode pops the top of
+the stack and evaluates it as a boolean argument.
+
+In the case of `IF-ELSE-FI` if the argument is true then the script between
+`IF` and `ELSE` is excuted followed by jumpting to the script after the `FI`.
+If it is a false then the script between the `ELSE` and the `FI` is executed.
+
+In the case of `IF-FI` if the argument is true then script between the `IF`
+and `FI` is executed. If it is false then flow jumps to the script after the
+`FI` opcode.
+
+## Encoding Formats
+
+The first version of CCLang supports the following encoding types:
+
+* Hexidecimal -- binary represented as lower case hexidecimal characters.
+* Base64 -- 62nd and 63rd characters are `+` and `/` respectively. No line
+  length maximum; all data in one line with mandatory padding using `=`.
+* Base64Url -- 62nd and 63rd characters are `-` and `_` respectively. No line
+  length maximum; all data in one line with no padding.
+* Base58 -- See definition [here](https://en.wikipedia.org/wiki/Base58).
+* Base58Check -- This is a complex construct and implemented using CCLang itself.
+
+## Encryption Algorithms
+
+The first version of CCLang supports the following encryption algorithms:
+
+* Curve25519
+* RSA
+* XSalsa20Poly1305
+* AES
+
+## Signing Algorithms
+
+The first version of CCLang supports the following signing algorithms:
+
+* Ed255519
+* ECDSA
+
+## Hashing Algorithms
+
+The first version of CCLang supports the following hashing algorithms:
+
+* SHA256
+* SHA512
+* SHA512-256
+* Blake2b
+* Argon2id
+
+## Serialization Formats
+
+CCLang is an abstract language definition and does not prescribe how the data
+and/or opcodes are serialized in any given format. Each encoding format is
+left to specify how each that is done. Below is a sample encoding specification
+for JSON.
+
+### JSON-CCLang
+
+The first job of mapping abstract CCLang to JSON is to decide the string
+constants to use for the supported encoding types, encryption algorithms,
+hashing algorithms, and opcodes. Below are lists of constants in CCLang and
+their string constant equivalents in JSON.
+
+#### Encoding Constants
+
+* Hexidecimal - `Hex`
+* Base64 - `Base64`
+* Base64Url - `Base64Url`
+* Base58 - `Base58`
+
+#### Encryption Algorithms
+
+* Curve25519 - `Curve25519`
+* RSA - `RSA`
+* XSalsa20Poly1305 - `XSalsa20Poly1305`
+* AES - `AES`
+
```

Comment:
Do we want to support any `ECIES` methods like libsodium's secretbox and sealedbox?

Reply:
XSalsa20Poly1305 is libsodium's SecretBox. In the v0.1.0 version I left out the SealedBox but if you look at the CCLang repo you'll see that it would be trivial to add new construct primitives.